clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -2964,6 +2964,100 @@ test_inactive()
+        make_instr(/*pc=*/101),
+        make_instr(/*pc=*/102),
+        make_timestamp(102),
+        make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+        make_marker(TRACE_MARKER_TYPE_SYSCALL, 999),
+        make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0),
+        make_marker(TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH, TID_C),
+        make_timestamp(401),
+        make_instr(/*pc=*/401),
+        make_exit(TID_A),
+    };
+    std::vector<trace_entry_t> refs_B = {
+        make_thread(TID_B),     make_pid(1),
+        make_timestamp(201),    make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+        make_instr(/*pc=*/201), make_instr(/*pc=*/202),
+        make_instr(/*pc=*/203), make_instr(/*pc=*/204),
+        make_instr(/*pc=*/205), make_instr(/*pc=*/206),
+        make_instr(/*pc=*/207), make_instr(/*pc=*/208),
+        make_instr(/*pc=*/209), make_instr(/*pc=*/210),
+        make_instr(/*pc=*/211), make_instr(/*pc=*/212),
+        make_instr(/*pc=*/213), make_instr(/*pc=*/214),
+        make_instr(/*pc=*/215), make_instr(/*pc=*/216),
+        make_instr(/*pc=*/217), make_instr(/*pc=*/218),
+        make_instr(/*pc=*/219), make_instr(/*pc=*/220),
+        make_exit(TID_B),
+    };
+    std::vector<trace_entry_t> refs_C = {
+        make_thread(TID_C),
+        make_pid(1),
+        make_timestamp(301),
+        make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+        make_instr(/*pc=*/301),
+        make_instr(/*pc=*/302),
+        make_timestamp(102),
+        make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+        make_marker(TRACE_MARKER_TYPE_SYSCALL, 999),
+        make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0),
+        make_marker(TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH, TID_A),
+        make_timestamp(501),
+        make_instr(/*pc=*/501),
+        // Test a non-existent target: should be ignored, but not crash.
+        make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0),
+        make_marker(TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH, TID_BASE + 3),
+        make_exit(TID_C),
+    };
+    std::vector<scheduler_t::input_reader_t> readers;
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_A)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_A);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_B)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_B);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_C)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_C);
+    // The lockstep treats markers like instructions so we have a long sequence
+    // of dots before we do the direct swap to C and back.
+    // TODO i#5843: Once we have i/o wait times we can make this more interesting
+    // by adding a D and having C be unschedule-able at switch time.
+    static const char *const CORE0_SCHED_STRING = "..AA........CC......A.C...";
```

Comment:
This schedule would be followed also if we had specified DEPENDENCY_TIMESTAMP right? But the test is verifying that it is followed even though we did not specify it?

Reply:
This is a good point.  I added a DEPENDENCY_DIRECT_SWITCH_BITFIELD and renamed DEPENDENCY_TIMESTAMPS to DEPENDENCY_TIMESTAMP_BITFIELD so we can combine them, and made a new enum entry DEPENDENCY_TIMESTAMPS which combines the two bitfields, which is what nearly every use case should want while still giving us control and without really breaking compatibility (and by providing bits and combinations the enum type is all that's needed still).  I changed this test to use the direct switch one.